### DISCUSSION TITLE
How does ScriptEngin.AccessContext work?

### DISCUSSION THREAD

**Discussion by nextfool** [ClearFoundry/ClearScript#328] [Q&A] [answered]
Hi @ClearScriptLib ,
So from my naïve understanding,  ScriptEngin.AccessContext can work as a "host api set" tool, by that I mean following:

> public class ApiSet1
{  
    public void Feature1(){}
    public void Feature2(){}
}
public class ApiSet2
{
    public void Feature1(){}
    public void Feature2(){}
}

 engineWithApiSet1.AccessContext = typeof(ApiSet1);
 engineWithApiSet2.AccessContext = typeof(ApiSet2);

But I was confused by the test methods in "AccessContextTest"

![image](https://user-images.githubusercontent.com/2167369/144568320-b10968b4-cd55-4367-a55f-a1c58bce51c9.png)

So here are my questions:
Given the fact that "AccessContextTestBase" is already "exposed" to engines global context by following,
![image](https://user-images.githubusercontent.com/2167369/144568456-2e80cf1f-395c-4e8f-90c5-cdde84ac8ee7.png)

How could code like " engine.Evaluate("new AccessContextTestBase()")" would fail?  I mean, the default engine.AccessContext  is the default "global" context isn't?

The next questions is that:
![image](https://user-images.githubusercontent.com/2167369/144568836-06d57be7-a782-4649-a1fa-a89fef037f53.png)
Neither of these two types defined a member "AccessContextTestBase" , then how suddenly after  engine.AccessContext = GetType(), those the same code can succeed?

Please help me to understand the magic here :)
Thank you for your time.

**Comment by nextfool** [comment]
Another question would be, how does "engine.AccessContext = typeof(SomeContext)" affect the default global context, are those types/objects added by "AddHostType/AddHostObject" still accessible/importable in script after the new context is assigned?

**Comment by ClearScriptLib** [comment, marked as answer]
Hi @nextfool,

>How could code like " engine.Evaluate("new AccessContextTestBase()")" would fail?

For script code to use a host resource, that resource must be both _exposed_ and _accessible_. As you've noted, `TestInitialize` exposes `AccessContextTestBase`. However, initially, that class isn't accessible, because it's internal to the ClearScript.Core assembly<sup>1</sup>.

>I mean, the default engine.AccessContext is the default "global" context isn't?

By default, `AccessContext` is `null`, which means that a host resource is accessible only if it has public visibility.

>Neither of these two types defined a member "AccessContextTestBase" , then how suddenly after engine.AccessContext = GetType(), those the same code can succeed?

Within that test method, `GetType` returns `typeof(AccessContextTest)`, so that statement effectively makes subsequent script code behave – in terms of resource accessibility – as if it were managed code within the `AccessContextTest` class.

In other words, that statement empowers script code to use the same resources as managed code within `AccessContextTest` – as long as those resources are exposed.

After that statement, `AccessContextTestBase` is not only exposed but also accessible, because script code now effectively resides within the ClearScriptTest assembly, and ClearScript.Core makes its internals visible to ClearScriptTest (see [here](https://github.com/microsoft/ClearScript/blob/d5f5baa17cc3a706c13697319c4b68fb4612f206/ClearScript/Properties/AssemblyInfo.Core.cs#L18)).

>Another question would be, how does "engine.AccessContext = typeof(SomeContext)" affect the default global context, are those types/objects added by "AddHostType/AddHostObject" still accessible/importable in script after the new context is assigned?

Again, `AccessContext` affects accessibility, whereas `AddHost*` affects exposure. Those concepts are independent.

Please send any additional questions our way.

Thank you!

<sup>1</sup> Yes, putting a test class into the core assembly is ugly, but it's the only way we could test certain things without creating yet another project.

**Reply by nextfool** [reply]
> Yes, putting a test class into the core assembly is ugly, but it's the only way we could test certain things without creating yet another project.

aha, it's actually such a brilliant solution, I couldn't imagine something else better. To be honest, I was never aware of the usage of "InternalsVisibleTo" attribute. So thank you very much for the clarification.
